### PR TITLE
Msudhir/add vector update functionality

### PR DESCRIPTION
**Pull request by Manisha4**
**What this PR does / why we need it**:

Added updated functionality to MilvusOnlineStore. The update functionality is able to delete and create a collection.
Relevant tests added as well

**Which issue(s) this PR fixes**:

Fixes #

### CODE REVIEW

File: sdk/python/feast/expediagroup/vectordb/milvus_online_store.py [hunk 1]
```diff
@@ -58,9 +98,34 @@ def update(
...
     ):
-        raise NotImplementedError(
-            "to be implemented in https://jira.expedia.biz/browse/EAPC-7970"
-        )
+        with MilvusConnectionManager(config.online_store):
+            for table_to_keep in tables_to_keep:
+                collection_available = utility.has_collection(table_to_keep.name)
+                try:
+                    if collection_available:
+                        logger.error(f"Collection {table_to_keep.name} already exists.")
```
Comment (michaelbackes) [inline sdk/python/feast/expediagroup/vectordb/milvus_online_store.py:106]: please make this log statement on "info" level. If a collection exists it is not an error and is expected.

File: sdk/python/tests/expediagroup/test_milvus_online_store.py [hunk 2]
```diff
@@ -1,37 +1,293 @@
...
+PROJECT = "test_aws"
+PROVIDER = "aws"
+TABLE_NAME = "milvus_online_store"
+REGION = "us-west-2"
+HOST = "localhost"
+ALIAS = "milvus"
+
+
+@dataclass
+class MockFeatureView:
```
Comment (michaelbackes) [inline sdk/python/tests/expediagroup/test_milvus_online_store.py:30]: is it really necessary to use a mock? Can you use a VectorFeatureView instead?

File: sdk/python/tests/expediagroup/test_milvus_online_store.py [hunk 3]
```diff
@@ -1,37 +1,293 @@
...
+
+        with MilvusConnectionManager(repo_config.online_store):
+            print("Doing something")
+
+        # Assert that connections.disconnect was called with the expected argument
+        mock_disconnect.assert_called_once_with(repo_config.online_store.alias)
+
+        # Assert that logger.info was called with expected messages
+        expected_log_calls = [
+            mocker.call("Closing the connection to Milvus"),
```
Comment (michaelbackes) [inline sdk/python/tests/expediagroup/test_milvus_online_store.py:102]: verification through logs is not a good practice in unit tests. Look into methods of "connections" like `get_connection_addr()` https://github.com/milvus-io/pymilvus/blob/master/pymilvus/orm/connections.py#L396 to use for the assert statements

File: sdk/python/tests/expediagroup/test_milvus_online_store.py [hunk 4]
```diff
@@ -1,37 +1,293 @@
...
-            "auto_id": True,
-        }
+            entities_to_delete=None,
+            entities_to_keep=None,
+            partial=None,
+        )
+        assert len(utility.list_collections()) == 1
+        assert utility.has_collection(self.collection_to_write) is True
+        assert (
+            f"Collection {self.collection_to_write} has been created successfully."
```
Comment (michaelbackes) [inline sdk/python/tests/expediagroup/test_milvus_online_store.py:157]: no need to assert a specific log statement here

File: sdk/python/tests/expediagroup/test_milvus_online_store.py [hunk 5]
```diff
@@ -1,37 +1,293 @@
...
+            tables_to_delete=[],
+            tables_to_keep=[
+                MockFeatureView(name=self.collection_to_write, schema=schema)
+            ],
+            entities_to_delete=None,
+            entities_to_keep=None,
+            partial=None,
+        )
+
+        assert f"Collection {self.collection_to_write} already exists." in caplog.text
```
Comment (michaelbackes) [inline sdk/python/tests/expediagroup/test_milvus_online_store.py:200]: assert with `list_collections() == 1` instead

File: sdk/python/tests/expediagroup/test_milvus_online_store.py [hunk 6]
```diff
@@ -1,37 +1,293 @@
...
+                    "int64", DataType.INT64, description="int64", is_primary=True
+                ),
+                FieldSchema(
+                    "float_vector", DataType.FLOAT_VECTOR, is_primary=False, dim=128
+                ),
+            ]
+        )
 
-    online_store_creator.teardown()
+        MilvusOnlineStore().update(
```
Comment (michaelbackes) [inline sdk/python/tests/expediagroup/test_milvus_online_store.py:177]: using the milvus online store here makes this test dependent on `update()` working as expected. Instead creating a collection via the pymilvus sdk directly would be better here to simulate that a collection already exists.

File: sdk/python/feast/expediagroup/vectordb/milvus_online_store.py [hunk 7]
```diff
@@ -58,9 +98,34 @@ def update(
...
+                    else:
+                        Collection(name=table_to_keep.name, schema=table_to_keep.schema)
```
Comment (michaelbackes) [inline sdk/python/feast/expediagroup/vectordb/milvus_online_store.py:108]: just catching this. FeatureView schemas are not the same as the schema that Milvus expects. We will need to write logic to translate that.

File: sdk/python/tests/expediagroup/test_milvus_online_store.py [hunk 8]
```diff
@@ -1,37 +1,293 @@
...
+                FieldSchema(
+                    "int64", DataType.INT64, description="int64", is_primary=True
+                ),
+                FieldSchema(
+                    "float_vector", DataType.FLOAT_VECTOR, is_primary=False, dim=128
+                ),
+            ]
+        )
+
+        MilvusOnlineStore().update(
```
Comment (michaelbackes) [inline sdk/python/tests/expediagroup/test_milvus_online_store.py:247]: here I would also rather create the collection first through the pymilvus sdk so that we do not introduce tests that are interdependent.

File: sdk/python/tests/expediagroup/test_milvus_online_store.py [hunk 9]
```diff
@@ -1,37 +1,293 @@
...
+            ],
+            tables_to_keep=[],
+            entities_to_delete=None,
+            entities_to_keep=None,
+            partial=None,
+        )
+        assert utility.has_collection(self.collection_to_delete) is False
+        assert len(utility.list_collections()) == 0
+        assert (
+            f"Collection {self.collection_to_delete} has been deleted successfully."
```
Comment (michaelbackes) [inline sdk/python/tests/expediagroup/test_milvus_online_store.py:271]: do not assert log-statements. If the wording changes one has to go and update the tests.

File: sdk/python/tests/expediagroup/test_milvus_online_store.py [hunk 10]
```diff
@@ -1,37 +1,293 @@
...
+                MockFeatureView(name=self.collection_to_delete, schema=None)
+            ],
+            tables_to_keep=[],
+            entities_to_delete=None,
+            entities_to_keep=None,
+            partial=None,
+        )
+
+        assert (
+            f"Collection {self.collection_to_delete} does not exist or is already deleted."
```
Comment (michaelbackes) [inline sdk/python/tests/expediagroup/test_milvus_online_store.py]: check with `assert connections.has_connection(self.collection_to_delete)` instead

File: sdk/python/feast/expediagroup/vectordb/milvus_online_store.py [hunk 11]
```diff
@@ -58,9 +116,39 @@ def update(
...
+            for table_to_delete in tables_to_delete:
+                collection_available = utility.has_collection(table_to_delete.name)
+                try:
+                    if collection_available:
+                        utility.drop_collection(table_to_delete.name)
+                        logger.info(
+                            f"Collection {table_to_delete.name} has been deleted successfully."
+                        )
+                    else:
+                        return logger.error(
```
Comment (michaelbackes) [inline sdk/python/feast/expediagroup/vectordb/milvus_online_store.py:147]: remove return statement

File: sdk/python/feast/expediagroup/vectordb/milvus_online_store.py [hunk 12]
```diff
@@ -71,3 +159,65 @@ def teardown(
...
+        """
+        boolean_mapping_from_string = {"True": True, "False": False}
+        field_list = []
+
+        for field in feast_schema:
+            data_type = self._feast_to_milvus_data_type(field.dtype)
+            field_name = field.name
+            description = field.tags.get("description")
+            is_primary = boolean_mapping_from_string.get(field.tags.get("is_primary"))
+            dimension = field.tags.get("dimension")
```
Comment (michaelbackes) [inline sdk/python/feast/expediagroup/vectordb/milvus_online_store.py:185]: dimensions have to be retrieved from `VectorFeatureView`. Also, the name of the column holding the vector is defined separately in `VectorFeatureView` ("vector_field")

File: sdk/python/feast/expediagroup/vectordb/milvus_online_store.py [hunk 13]
```diff
@@ -71,3 +159,65 @@ def teardown(
...
+
+            if dimension is not None:
```
Comment (michaelbackes) [inline sdk/python/feast/expediagroup/vectordb/milvus_online_store.py:187]: why is this needed?

File: sdk/python/feast/expediagroup/vectordb/vector_feature_view.py [hunk 14]
```diff
@@ -50,6 +50,7 @@ class VectorFeatureView(BaseFeatureView):
 
     # inheriting from FeatureView wouldn't work due to issue with conflicting proto classes
     # therefore using composition instead
+    name: str
```
Comment (michaelbackes) [inline sdk/python/feast/expediagroup/vectordb/vector_feature_view.py:53]: feature_view already has an attribute `name`

File: sdk/python/tests/expediagroup/test_milvus_online_store.py [hunk 15]
```diff
@@ -1,37 +1,346 @@
...
+
+
+class TestMilvusOnlineStore:
+
+    collection_to_write = "Collection2"
+    collection_to_delete = "Collection1"
+
+    def setup_method(self, milvus_online_setup):
+        # Ensuring that the collections created are dropped before the tests are run
+        with PymilvusConnectionContext():
```
Comment (michaelbackes) [inline sdk/python/tests/expediagroup/test_milvus_online_store.py:131]: it's ok to re-use `MilvusConnectionManager` in order to not duplicate code

File: sdk/python/tests/expediagroup/test_milvus_online_store.py [hunk 16]
```diff
@@ -1,37 +1,346 @@
...
+                dtype=Int64,
+                tags={"is_primary": "True", "description": "int64"},
+            ),
+            Field(
+                name="feature1",
+                dtype=Array(Float32),
+                tags={
+                    "is_primary": "False",
+                    "description": "float32",
+                    "dimension": "128",
```
Comment (michaelbackes) [inline sdk/python/tests/expediagroup/test_milvus_online_store.py:155]: thinking about it this is actually a great suggestion to use a tag to pass along the number of dimensions. I think we can do likewise with index algorithm and eliminate the need for a `VectorFeatureView` altogether. What do you think?

File: sdk/python/tests/expediagroup/test_milvus_online_store.py [hunk 17]
```diff
@@ -1,37 +1,346 @@
...
+                    index_algorithm=INDEX_ALGO,
+                )
+            ],
+            entities_to_delete=None,
+            entities_to_keep=None,
+            partial=None,
+        )
+
+        # Here we want to open and add a collection using pymilvus directly and close the connection, we need to check if the collection count remains 1 and exists.
+        with PymilvusConnectionContext():
```
Comment (michaelbackes) [inline sdk/python/tests/expediagroup/test_milvus_online_store.py:243]: does the embedded Milvus actually persist data after it is stopped?

Comment (Manisha4) [reply sdk/python/tests/expediagroup/test_milvus_online_store.py:243]: Yes it does

File: sdk/python/tests/expediagroup/test_milvus_online_store.py [hunk 18]
```diff
@@ -1,37 +1,346 @@
...
+                    index_algorithm=INDEX_ALGO,
+                )
+            ],
+            tables_to_keep=[],
+            entities_to_delete=None,
+            entities_to_keep=None,
+            partial=None,
+        )
+
+        assert "Collection abc does not exist or is already deleted." in caplog.text
```
Comment (michaelbackes) [inline sdk/python/tests/expediagroup/test_milvus_online_store.py:346]: minor: I'd rather assert against Milvus not having any collection. The update() method should really have a return value, but that is an issue in Feast itself

File: sdk/python/feast/expediagroup/vectordb/vector_feature_view.py (hunk 14 above)
Comment (Manisha4) [reply sdk/python/feast/expediagroup/vectordb/vector_feature_view.py:53]: It did not reflect and threw an error when I tried to add name

File: sdk/python/tests/expediagroup/test_milvus_online_store.py (hunk 16 above)
Comment (Manisha4) [reply sdk/python/tests/expediagroup/test_milvus_online_store.py]: Yes, thats a good idea

File: sdk/python/tests/expediagroup/test_milvus_online_store.py (hunk 15 above)
Comment (Manisha4) [reply sdk/python/tests/expediagroup/test_milvus_online_store.py:131]: I want to keep the test not dependent on any functionality in Milvus online store. This context manager is a very simple open and close connection, purely for testing purposes. Will not add in username, password, etc.

**Review by michaelbackes — APPROVED** [review]
great work!
